From ead41e71bcfe73bfb27331753857578dd561cb58 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 28 Mar 2021 20:38:00 -0500
Subject: [PATCH 060/196] Thread ScriptExecutionData to SignatureHash. (#112)

ScriptExecutionData contains codeseparator_pos and executed_script_hash, which is required for the BIP341 sighash.

Co-authored-by: Shammah Chancellor <schancel@users.noreply.github.com>
---
 src/bench/verify_script.cpp          |  2 +-
 src/script/interpreter.cpp           | 87 +++++++++++++++++-----------
 src/script/interpreter.h             | 11 ++--
 src/script/script_exec_data.h        | 33 ++++++++++-
 src/script/sign.cpp                  | 24 +++++---
 src/script/sign.h                    |  2 +
 src/test/fuzz/script_sign.cpp        |  3 +-
 src/test/multisig_tests.cpp          |  6 +-
 src/test/script_tests.cpp            | 29 ++++++----
 src/test/sigcheckcount_tests.cpp     |  5 +-
 src/test/sighash_bip341_tests.cpp    |  2 +-
 src/test/sighash_tests.cpp           | 55 ++++++++++--------
 src/test/txvalidationcache_tests.cpp | 29 ++++++----
 13 files changed, 187 insertions(+), 101 deletions(-)

diff --git a/src/bench/verify_script.cpp b/src/bench/verify_script.cpp
index 6dc72d2e6..12886cca4 100644
--- a/src/bench/verify_script.cpp
+++ b/src/bench/verify_script.cpp
@@ -31,7 +31,7 @@ static void VerifyNestedIfScript(benchmark::Bench &bench) {
     bench.run([&] {
         auto stack_copy = stack;
         ScriptExecutionMetrics metrics = {};
-        ScriptExecutionData execdata = {};
+        ScriptExecutionData execdata{CScript()};
         ScriptError error;
         bool ret = EvalScript(stack_copy, script, 0, BaseSignatureChecker(),
                               metrics, execdata, &error);
diff --git a/src/script/interpreter.cpp b/src/script/interpreter.cpp
index 14967c71c..c2eb49cf5 100644
--- a/src/script/interpreter.cpp
+++ b/src/script/interpreter.cpp
@@ -186,8 +186,9 @@ static bool EvalChecksig(const valtype &vchSig, const valtype &vchPubKey,
                          CScript::const_iterator pbegincodehash,
                          CScript::const_iterator pend, uint32_t flags,
                          const BaseSignatureChecker &checker,
-                         ScriptExecutionMetrics &metrics, ScriptError *serror,
-                         bool &fSuccess) {
+                         ScriptExecutionMetrics &metrics,
+                         const std::optional<ScriptExecutionData> &execdata,
+                         ScriptError *serror, bool &fSuccess) {
     if (!CheckTransactionSignatureEncoding(vchSig, flags, serror) ||
         !CheckPubKeyEncoding(vchPubKey, flags, serror)) {
         // serror is set
@@ -202,7 +203,8 @@ static bool EvalChecksig(const valtype &vchSig, const valtype &vchPubKey,
         // Remove signature for pre-fork scripts
         CleanupScriptCode(scriptCode, vchSig, flags);
 
-        fSuccess = checker.CheckSig(vchSig, vchPubKey, scriptCode, flags);
+        fSuccess =
+            checker.CheckSig(vchSig, vchPubKey, execdata, scriptCode, flags);
         metrics.nSigChecks += 1;
 
         if (!fSuccess) {
@@ -944,7 +946,8 @@ bool EvalScript(std::vector<valtype> &stack, const CScript &script,
 
                         bool fSuccess = false;
                         if (!EvalChecksig(vchSig, vchPubKey, pbegincodehash,
-                                          pend, flags, checker, metrics, serror,
+                                          pend, flags, checker, metrics,
+                                          std::optional(execdata), serror,
                                           fSuccess)) {
                             return false;
                         }
@@ -1134,6 +1137,7 @@ bool EvalScript(std::vector<valtype> &stack, const CScript &script,
 
                                 // Check signature
                                 if (!checker.CheckSig(vchSig, vchPubKey,
+                                                      std::optional(execdata),
                                                       scriptCode, flags)) {
                                     // This can fail if the signature is empty,
                                     // which also is a NULLFAIL error as the
@@ -1185,8 +1189,9 @@ bool EvalScript(std::vector<valtype> &stack, const CScript &script,
                                 }
 
                                 // Check signature
-                                bool fOk = checker.CheckSig(vchSig, vchPubKey,
-                                                            scriptCode, flags);
+                                bool fOk = checker.CheckSig(
+                                    vchSig, vchPubKey, std::optional(execdata),
+                                    scriptCode, flags);
 
                                 if (fOk) {
                                     nSigsRemaining--;
@@ -1772,10 +1777,11 @@ bool SignatureHashBIP143(uint256 &sighashOut, const CScript &scriptCode,
 }
 
 template <class T>
-bool SignatureHash(uint256 &sighashOut, const CScript &scriptCode,
-                   const T &txTo, unsigned int nIn, SigHashType sigHashType,
-                   const Amount amount, const PrecomputedTransactionData *cache,
-                   uint32_t flags) {
+bool SignatureHash(uint256 &sighashOut,
+                   const std::optional<ScriptExecutionData> &execdata,
+                   const CScript &scriptCode, const T &txTo, unsigned int nIn,
+                   SigHashType sigHashType, const Amount amount,
+                   const PrecomputedTransactionData *cache, uint32_t flags) {
     assert(nIn < txTo.vin.size());
 
     if (flags & SCRIPT_ENABLE_REPLAY_PROTECTION) {
@@ -1796,16 +1802,18 @@ bool SignatureHash(uint256 &sighashOut, const CScript &scriptCode,
 }
 
 // need to instantiate explicitly
-template bool SignatureHash(uint256 &sighashOut, const CScript &scriptCode,
-                            const CTransaction &txTo, unsigned int nIn,
-                            SigHashType sigHashType, const Amount amount,
-                            const PrecomputedTransactionData *cache,
-                            uint32_t flags);
-template bool SignatureHash(uint256 &sighashOut, const CScript &scriptCode,
-                            const CMutableTransaction &txTo, unsigned int nIn,
-                            SigHashType sigHashType, const Amount amount,
-                            const PrecomputedTransactionData *cache,
-                            uint32_t flags);
+template bool
+SignatureHash(uint256 &sighashOut,
+              const std::optional<ScriptExecutionData> &execdata,
+              const CScript &scriptCode, const CTransaction &txTo,
+              unsigned int nIn, SigHashType sigHashType, const Amount amount,
+              const PrecomputedTransactionData *cache, uint32_t flags);
+template bool
+SignatureHash(uint256 &sighashOut,
+              const std::optional<ScriptExecutionData> &execdata,
+              const CScript &scriptCode, const CMutableTransaction &txTo,
+              unsigned int nIn, SigHashType sigHashType, const Amount amount,
+              const PrecomputedTransactionData *cache, uint32_t flags);
 
 bool BaseSignatureChecker::VerifySignature(const std::vector<uint8_t> &vchSig,
                                            const CPubKey &pubkey,
@@ -1820,6 +1828,7 @@ bool BaseSignatureChecker::VerifySignature(const std::vector<uint8_t> &vchSig,
 template <class T>
 bool GenericTransactionSignatureChecker<T>::CheckSig(
     const std::vector<uint8_t> &vchSigIn, const std::vector<uint8_t> &vchPubKey,
+    const std::optional<ScriptExecutionData> &execdata,
     const CScript &scriptCode, uint32_t flags) const {
     CPubKey pubkey(vchPubKey);
     if (!pubkey.IsValid()) {
@@ -1835,8 +1844,8 @@ bool GenericTransactionSignatureChecker<T>::CheckSig(
     vchSig.pop_back();
 
     uint256 sighash;
-    if (!SignatureHash(sighash, scriptCode, *txTo, nIn, sigHashType, amount,
-                       this->txdata, flags)) {
+    if (!SignatureHash(sighash, execdata, scriptCode, *txTo, nIn, sigHashType,
+                       amount, this->txdata, flags)) {
         return false;
     }
 
@@ -1951,21 +1960,26 @@ bool VerifyScript(const CScript &scriptSig, const CScript &scriptPubKey,
     }
 
     ScriptExecutionMetrics metrics = {};
-    ScriptExecutionData execdata = {};
 
     // scriptSig and scriptPubKey must be evaluated sequentially on the same
     // stack rather than being simply concatenated (see CVE-2010-5141)
     std::vector<valtype> stack, stackCopy;
-    if (!EvalScript(stack, scriptSig, flags, checker, metrics, execdata,
-                    serror)) {
-        // serror is set
-        return false;
+    {
+        ScriptExecutionData execdata{CScript()};  // unused in scriptSig
+        if (!EvalScript(stack, scriptSig, flags, checker, metrics, execdata,
+                        serror)) {
+            // serror is set
+            return false;
+        }
     }
     stackCopy = stack;
-    if (!EvalScript(stack, scriptPubKey, flags, checker, metrics, execdata,
-                    serror)) {
-        // serror is set
-        return false;
+    {
+        ScriptExecutionData execdata{scriptPubKey};
+        if (!EvalScript(stack, scriptPubKey, flags, checker, metrics, execdata,
+                        serror)) {
+            // serror is set
+            return false;
+        }
     }
     if (stack.empty()) {
         return set_error(serror, ScriptError::EVAL_FALSE);
@@ -1993,10 +2007,13 @@ bool VerifyScript(const CScript &scriptSig, const CScript &scriptPubKey,
         CScript pubKey2(pubKeySerialized.begin(), pubKeySerialized.end());
         popstack(stack);
 
-        if (!EvalScript(stack, pubKey2, flags, checker, metrics, execdata,
-                        serror)) {
-            // serror is set
-            return false;
+        {
+            ScriptExecutionData execdata{pubKey2};
+            if (!EvalScript(stack, pubKey2, flags, checker, metrics, execdata,
+                            serror)) {
+                // serror is set
+                return false;
+            }
         }
         if (stack.empty()) {
             return set_error(serror, ScriptError::EVAL_FALSE);
diff --git a/src/script/interpreter.h b/src/script/interpreter.h
index aa17010e1..3407252d8 100644
--- a/src/script/interpreter.h
+++ b/src/script/interpreter.h
@@ -23,9 +23,10 @@ class CTransaction;
 class uint256;
 
 template <class T>
-bool SignatureHash(uint256 &sighashOut, const CScript &scriptCode,
-                   const T &txTo, unsigned int nIn, SigHashType sigHashType,
-                   const Amount amount,
+bool SignatureHash(uint256 &sighashOut,
+                   const std::optional<ScriptExecutionData> &execdata,
+                   const CScript &scriptCode, const T &txTo, unsigned int nIn,
+                   SigHashType sigHashType, const Amount amount,
                    const PrecomputedTransactionData *cache = nullptr,
                    uint32_t flags = SCRIPT_ENABLE_SIGHASH_FORKID);
 
@@ -37,6 +38,7 @@ public:
 
     virtual bool CheckSig(const std::vector<uint8_t> &vchSigIn,
                           const std::vector<uint8_t> &vchPubKey,
+                          const std::optional<ScriptExecutionData> &execdata,
                           const CScript &scriptCode, uint32_t flags) const {
         return false;
     }
@@ -69,6 +71,7 @@ public:
     // The overridden functions are now final.
     bool CheckSig(const std::vector<uint8_t> &vchSigIn,
                   const std::vector<uint8_t> &vchPubKey,
+                  const std::optional<ScriptExecutionData> &execdata,
                   const CScript &scriptCode,
                   uint32_t flags) const final override;
     bool CheckLockTime(const CScriptNum &nLockTime) const final override;
@@ -89,7 +92,7 @@ static inline bool EvalScript(std::vector<std::vector<uint8_t>> &stack,
                               const BaseSignatureChecker &checker,
                               ScriptError *error = nullptr) {
     ScriptExecutionMetrics dummymetrics;
-    ScriptExecutionData dummyexecdata;
+    ScriptExecutionData dummyexecdata{script};
     return EvalScript(stack, script, flags, checker, dummymetrics,
                       dummyexecdata, error);
 }
diff --git a/src/script/script_exec_data.h b/src/script/script_exec_data.h
index c7fe847e6..eff6919d7 100644
--- a/src/script/script_exec_data.h
+++ b/src/script/script_exec_data.h
@@ -5,6 +5,8 @@
 #ifndef BITCOIN_SCRIPT_SCRIPT_EXEC_DATA_H
 #define BITCOIN_SCRIPT_SCRIPT_EXEC_DATA_H
 
+#include <hash.h>
+#include <script/script.h>
 #include <uint256.h>
 
 /**
@@ -12,12 +14,41 @@
  * during script execution, e.g. for computing sig hashes.
  */
 struct ScriptExecutionData {
+    static constexpr uint32_t DEFAULT_CODESEP_POS = 0xffff'ffff;
+
     /**
      * Opcode position of the last executed OP_CODESEPARATOR.
-     * 
+     *
      * This allows signatures to commit to certain code paths.
      */
     uint32_t m_codeseparator_pos;
+
+    /**
+     * SHA256 of the (complete) script being executed, regardless of
+     * OP_CODESEPARATOR.
+     *
+     * - In plain scripts, this is the SHA256 of the scriptPubKey.
+     * - In P2SH, this is the SHA256 of the redeemScript.
+     * - In Tapscripts, this is the "tapleaf hash", calculated as defined in
+     *   BIP341: taggedhash_TapLeaf(version || compact_size(len(script)) ||
+     *                              script).
+     * - In Taproot, this entire struct would not be constructed.
+     */
+    uint256 m_executed_script_hash;
+
+    /**
+     * ScriptExecutionData with the SHA-256 of executed_script and as
+     * codeseparator position.
+     */
+    ScriptExecutionData(const CScript &executed_script,
+                        uint32_t codeseparator_pos = DEFAULT_CODESEP_POS) {
+        CSHA256()
+            .Write(executed_script.data(), executed_script.size())
+            .Finalize(m_executed_script_hash.begin());
+        m_codeseparator_pos = codeseparator_pos;
+    }
+
+    ScriptExecutionData() = delete;
 };
 
 #endif // BITCOIN_SCRIPT_SCRIPT_EXEC_DATA_H
diff --git a/src/script/sign.cpp b/src/script/sign.cpp
index a9c4e533b..270e3922c 100644
--- a/src/script/sign.cpp
+++ b/src/script/sign.cpp
@@ -23,13 +23,15 @@ MutableTransactionSignatureCreator::MutableTransactionSignatureCreator(
 
 bool MutableTransactionSignatureCreator::CreateSig(
     const SigningProvider &provider, std::vector<uint8_t> &vchSig,
-    const CKeyID &address, const CScript &scriptCode) const {
+    const CKeyID &address, const std::optional<ScriptExecutionData> &execdata,
+    const CScript &scriptCode) const {
     CKey key;
     if (!provider.GetKey(address, key)) {
         return false;
     }
     uint256 hash;
-    if (!SignatureHash(hash, scriptCode, *txTo, nIn, sigHashType, amount)) {
+    if (!SignatureHash(hash, execdata, scriptCode, *txTo, nIn, sigHashType,
+                       amount)) {
         return false;
     }
     if (!key.SignECDSA(hash, vchSig)) {
@@ -76,7 +78,7 @@ static bool GetPubKey(const SigningProvider &provider,
 static bool CreateSig(const BaseSignatureCreator &creator,
                       SignatureData &sigdata, const SigningProvider &provider,
                       std::vector<uint8_t> &sig_out, const CPubKey &pubkey,
-                      const CScript &scriptcode) {
+                      const CScript &scriptPubKey) {
     CKeyID keyid = pubkey.GetID();
     const auto it = sigdata.signatures.find(keyid);
     if (it != sigdata.signatures.end()) {
@@ -88,7 +90,9 @@ static bool CreateSig(const BaseSignatureCreator &creator,
         sigdata.misc_pubkeys.emplace(keyid,
                                      std::make_pair(pubkey, std::move(info)));
     }
-    if (creator.CreateSig(provider, sig_out, keyid, scriptcode)) {
+    if (creator.CreateSig(provider, sig_out, keyid,
+                          ScriptExecutionData(scriptPubKey),
+                          scriptPubKey)) {
         auto i = sigdata.signatures.emplace(keyid, SigPair(pubkey, sig_out));
         assert(i.second);
         return true;
@@ -246,8 +250,10 @@ public:
         : sigdata(sigdata_), checker(checker_) {}
     bool CheckSig(const std::vector<uint8_t> &scriptSig,
                   const std::vector<uint8_t> &vchPubKey,
+                  const std::optional<ScriptExecutionData> &execdata,
                   const CScript &scriptCode, uint32_t flags) const override {
-        if (checker.CheckSig(scriptSig, vchPubKey, scriptCode, flags)) {
+        if (checker.CheckSig(scriptSig, vchPubKey, execdata, scriptCode,
+                             flags)) {
             CPubKey pubkey(vchPubKey);
 
             sigdata.signatures.emplace(pubkey.GetID(),
@@ -321,8 +327,10 @@ SignatureData DataFromTransaction(const CMutableTransaction &tx,
                 // We either have a signature for this pubkey, or we have found
                 // a signature and it is valid
                 if (data.signatures.count(CPubKey(pubkey).GetID()) ||
-                    extractor_checker.CheckSig(sig, pubkey, next_script,
-                                               STANDARD_SCRIPT_VERIFY_FLAGS)) {
+                    extractor_checker.CheckSig(
+                        sig, pubkey,
+                        ScriptExecutionData(next_script),
+                        next_script, STANDARD_SCRIPT_VERIFY_FLAGS)) {
                     last_success_key = i + 1;
                     break;
                 }
@@ -396,6 +404,7 @@ public:
     DummySignatureChecker() {}
     bool CheckSig(const std::vector<uint8_t> &scriptSig,
                   const std::vector<uint8_t> &vchPubKey,
+                  const std::optional<ScriptExecutionData> &execdata,
                   const CScript &scriptCode, uint32_t flags) const override {
         return true;
     }
@@ -415,6 +424,7 @@ public:
     }
     bool CreateSig(const SigningProvider &provider,
                    std::vector<uint8_t> &vchSig, const CKeyID &keyid,
+                   const std::optional<ScriptExecutionData> &execdata,
                    const CScript &scriptCode) const override {
         // Create a dummy signature that is a valid DER-encoding
         vchSig.assign(m_r_len + m_s_len + 7, '\000');
diff --git a/src/script/sign.h b/src/script/sign.h
index 830729208..6692983a3 100644
--- a/src/script/sign.h
+++ b/src/script/sign.h
@@ -31,6 +31,7 @@ public:
     /** Create a singular (non-script) signature. */
     virtual bool CreateSig(const SigningProvider &provider,
                            std::vector<uint8_t> &vchSig, const CKeyID &keyid,
+                           const std::optional<ScriptExecutionData> &execdata,
                            const CScript &scriptCode) const = 0;
 };
 
@@ -50,6 +51,7 @@ public:
     const BaseSignatureChecker &Checker() const override { return checker; }
     bool CreateSig(const SigningProvider &provider,
                    std::vector<uint8_t> &vchSig, const CKeyID &keyid,
+                   const std::optional<ScriptExecutionData> &execdata,
                    const CScript &scriptCode) const override;
 };
 
diff --git a/src/test/fuzz/script_sign.cpp b/src/test/fuzz/script_sign.cpp
index 2b2e8288e..c3cca2bc0 100644
--- a/src/test/fuzz/script_sign.cpp
+++ b/src/test/fuzz/script_sign.cpp
@@ -146,8 +146,9 @@ void test_one_input(const std::vector<uint8_t> &buffer) {
                 } else {
                     address = CKeyID{ConsumeUInt160(fuzzed_data_provider)};
                 }
+                const ScriptExecutionData execdata{CScript()};
                 (void)signature_creator.CreateSig(
-                    provider, vch_sig, address,
+                    provider, vch_sig, address, std::optional(execdata),
                     ConsumeScript(fuzzed_data_provider));
             }
             std::map<COutPoint, Coin> coins;
diff --git a/src/test/multisig_tests.cpp b/src/test/multisig_tests.cpp
index dee2290ad..f9bedb9fa 100644
--- a/src/test/multisig_tests.cpp
+++ b/src/test/multisig_tests.cpp
@@ -23,8 +23,10 @@ static CScript sign_multisig(const CScript &scriptPubKey,
                              const std::vector<CKey> &keys,
                              const CMutableTransaction &tx, int whichIn) {
     uint256 hash;
-    BOOST_CHECK(SignatureHash(hash, scriptPubKey, CTransaction(tx), whichIn,
-                              SigHashType(), Amount::zero()));
+    BOOST_CHECK(SignatureHash(
+        hash, std::optional(ScriptExecutionData(scriptPubKey)),
+        scriptPubKey, CTransaction(tx), whichIn, SigHashType(),
+        Amount::zero()));
 
     CScript result;
     // CHECKMULTISIG bug workaround
diff --git a/src/test/script_tests.cpp b/src/test/script_tests.cpp
index ecccf21ae..3f869e5d6 100644
--- a/src/test/script_tests.cpp
+++ b/src/test/script_tests.cpp
@@ -341,8 +341,9 @@ public:
                  Amount amount = Amount::zero(),
                  uint32_t sigFlags = SCRIPT_ENABLE_SIGHASH_FORKID) {
         uint256 hash;
-        BOOST_CHECK(SignatureHash(hash, script, CTransaction(spendTx), 0,
-                                  sigHashType, amount, nullptr, sigFlags));
+        BOOST_CHECK(SignatureHash(
+            hash, std::optional(ScriptExecutionData(script)), script,
+            CTransaction(spendTx), 0, sigHashType, amount, nullptr, sigFlags));
         std::vector<uint8_t> vchSig = DoSignECDSA(key, hash, lenR, lenS);
         vchSig.push_back(static_cast<uint8_t>(sigHashType.getRawSigHashType()));
         DoPush(vchSig);
@@ -354,8 +355,9 @@ public:
                    Amount amount = Amount::zero(),
                    uint32_t sigFlags = SCRIPT_ENABLE_SIGHASH_FORKID) {
         uint256 hash;
-        BOOST_CHECK(SignatureHash(hash, script, CTransaction(spendTx), 0,
-                                  sigHashType, amount, nullptr, sigFlags));
+        BOOST_CHECK(SignatureHash(
+            hash, std::optional(ScriptExecutionData(script)), script,
+            CTransaction(spendTx), 0, sigHashType, amount, nullptr, sigFlags));
         std::vector<uint8_t> vchSig = DoSignSchnorr(key, hash);
         vchSig.push_back(static_cast<uint8_t>(sigHashType.getRawSigHashType()));
         DoPush(vchSig);
@@ -389,8 +391,9 @@ public:
         // This calculates a pubkey to verify with a given ECDSA transaction
         // signature.
         uint256 hash;
-        BOOST_CHECK(SignatureHash(hash, script, CTransaction(spendTx), 0,
-                                  sigHashType, amount, nullptr, sigFlags));
+        BOOST_CHECK(SignatureHash(hash, ScriptExecutionData(script), script,
+                                  CTransaction(spendTx), 0, sigHashType, amount,
+                                  nullptr, sigFlags));
 
         assert(rdata.size() <= 32);
         assert(sdata.size() <= 32);
@@ -2157,8 +2160,9 @@ static CScript sign_multisig(const CScript &scriptPubKey,
                              const std::vector<CKey> &keys,
                              const CTransaction &transaction) {
     uint256 hash;
-    BOOST_CHECK(SignatureHash(hash, scriptPubKey, transaction, 0, SigHashType(),
-                              Amount::zero()));
+    BOOST_CHECK(SignatureHash(
+        hash, std::optional(ScriptExecutionData(scriptPubKey)), scriptPubKey,
+        transaction, 0, SigHashType(), Amount::zero()));
 
     CScript result;
     //
@@ -2456,14 +2460,16 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     // A couple of partially-signed versions:
     std::vector<uint8_t> sig1;
     uint256 hash1;
-    BOOST_CHECK(SignatureHash(hash1, scriptPubKey, CTransaction(txTo), 0,
+    BOOST_CHECK(SignatureHash(hash1, ScriptExecutionData(scriptPubKey),
+                              scriptPubKey, CTransaction(txTo), 0,
                               SigHashType().withForkId(), Amount::zero()));
     BOOST_CHECK(keys[0].SignECDSA(hash1, sig1));
     sig1.push_back(SIGHASH_ALL | SIGHASH_FORKID);
     std::vector<uint8_t> sig2;
     uint256 hash2;
     BOOST_CHECK(SignatureHash(
-        hash2, scriptPubKey, CTransaction(txTo), 0,
+        hash2, ScriptExecutionData(scriptPubKey), scriptPubKey,
+        CTransaction(txTo), 0,
         SigHashType().withBaseType(BaseSigHashType::NONE).withForkId(),
         Amount::zero()));
     BOOST_CHECK(keys[1].SignECDSA(hash2, sig2));
@@ -2471,7 +2477,8 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     std::vector<uint8_t> sig3;
     uint256 hash3;
     BOOST_CHECK(SignatureHash(
-        hash3, scriptPubKey, CTransaction(txTo), 0,
+        hash3, ScriptExecutionData(scriptPubKey), scriptPubKey,
+        CTransaction(txTo), 0,
         SigHashType().withBaseType(BaseSigHashType::SINGLE).withForkId(),
         Amount::zero()));
     BOOST_CHECK(keys[2].SignECDSA(hash3, sig3));
diff --git a/src/test/sigcheckcount_tests.cpp b/src/test/sigcheckcount_tests.cpp
index b92559e99..1a65a86ea 100644
--- a/src/test/sigcheckcount_tests.cpp
+++ b/src/test/sigcheckcount_tests.cpp
@@ -72,6 +72,7 @@ static class : public BaseSignatureChecker {
 
     bool CheckSig(const std::vector<uint8_t> &vchSigIn,
                   const std::vector<uint8_t> &vchPubKey,
+                  const std::optional<ScriptExecutionData> &execdata,
                   const CScript &scriptCode, uint32_t flags) const final {
         if (vchPubKey == badpub) {
             return false;
@@ -114,7 +115,7 @@ static void CheckEvalScript(const stacktype &original_stack,
         ScriptError err = ScriptError::UNKNOWN;
         stacktype stack{original_stack};
         ScriptExecutionMetrics metrics;
-        ScriptExecutionData execdata;
+        ScriptExecutionData execdata{script};
 
         bool r = EvalScript(stack, script, flags, dummysigchecker, metrics,
                             execdata, &err);
@@ -212,7 +213,7 @@ BOOST_AUTO_TEST_CASE(test_evalscript) {
     {
         stacktype stack{txsigschnorr};
         ScriptExecutionMetrics metrics;
-        ScriptExecutionData execdata;
+        ScriptExecutionData execdata{CScript()};
         metrics.nSigChecks = 12345;
         bool r = EvalScript(stack, CScript() << pub << OP_CHECKSIG,
                             SCRIPT_ENABLE_SIGHASH_FORKID, dummysigchecker,
diff --git a/src/test/sighash_bip341_tests.cpp b/src/test/sighash_bip341_tests.cpp
index 8241c1815..9fa433012 100644
--- a/src/test/sighash_bip341_tests.cpp
+++ b/src/test/sighash_bip341_tests.cpp
@@ -113,7 +113,7 @@ void CheckCodesepPos(const CScript &script,
     for (uint32_t flags : allflags) {
         BaseSignatureChecker sigchecker;
         ScriptExecutionMetrics metrics = {};
-        ScriptExecutionData execdata = {};
+        ScriptExecutionData execdata{CScript()};
         stacktype stack = {};
         bool r =
             EvalScript(stack, script, flags, sigchecker, metrics, execdata);
diff --git a/src/test/sighash_tests.cpp b/src/test/sighash_tests.cpp
index ebc15719c..410da96ad 100644
--- a/src/test/sighash_tests.cpp
+++ b/src/test/sighash_tests.cpp
@@ -138,19 +138,23 @@ BOOST_AUTO_TEST_CASE(sighash_test) {
         RandomTransaction(txTo, (nHashType & 0x1f) == SIGHASH_SINGLE);
         CScript scriptCode;
         RandomScript(scriptCode);
+        const std::optional<ScriptExecutionData> execdata{
+            ScriptExecutionData(scriptCode)};
         int nIn = InsecureRandRange(txTo.vin.size());
 
         uint256 shref =
             SignatureHashOld(scriptCode, CTransaction(txTo), nIn, nHashType);
         uint256 shold;
-        BOOST_CHECK(SignatureHash(shold, scriptCode, CTransaction(txTo), nIn,
-                                  sigHashType, Amount::zero(), nullptr, 0));
+        BOOST_CHECK(SignatureHash(shold, execdata, scriptCode,
+                                  CTransaction(txTo), nIn, sigHashType,
+                                  Amount::zero(), nullptr, 0));
         BOOST_CHECK(shold == shref);
 
         // Check the impact of the forkid flag.
         uint256 shreg;
-        BOOST_CHECK(SignatureHash(shreg, scriptCode, CTransaction(txTo), nIn,
-                                  sigHashType, Amount::zero()));
+        BOOST_CHECK(SignatureHash(shreg, execdata, scriptCode,
+                                  CTransaction(txTo), nIn, sigHashType,
+                                  Amount::zero()));
         if (sigHashType.hasForkId()) {
             BOOST_CHECK(nHashType & SIGHASH_FORKID);
             BOOST_CHECK(shreg != shref);
@@ -161,38 +165,38 @@ BOOST_AUTO_TEST_CASE(sighash_test) {
 
         // Make sure replay protection works as expected.
         uint256 shrep;
-        BOOST_CHECK(SignatureHash(shrep, scriptCode, CTransaction(txTo), nIn,
-                                  sigHashType, Amount::zero(), nullptr,
-                                  SCRIPT_ENABLE_SIGHASH_FORKID |
-                                      SCRIPT_ENABLE_REPLAY_PROTECTION));
+        BOOST_CHECK(SignatureHash(
+            shrep, execdata, scriptCode, CTransaction(txTo), nIn, sigHashType,
+            Amount::zero(), nullptr,
+            SCRIPT_ENABLE_SIGHASH_FORKID | SCRIPT_ENABLE_REPLAY_PROTECTION));
         uint32_t newForkValue = 0xff0000 | ((nHashType >> 8) ^ 0xdead);
         uint256 manualshrep;
-        BOOST_CHECK(SignatureHash(manualshrep, scriptCode, CTransaction(txTo),
-                                  nIn, sigHashType.withForkValue(newForkValue),
-                                  Amount::zero()));
+        BOOST_CHECK(SignatureHash(
+            manualshrep, execdata, scriptCode, CTransaction(txTo), nIn,
+            sigHashType.withForkValue(newForkValue), Amount::zero()));
         BOOST_CHECK(shrep == manualshrep);
 
         // Replay protection works even if the hash is of the form 0xffxxxx
         uint256 shrepff;
         BOOST_CHECK(SignatureHash(
-            shrepff, scriptCode, CTransaction(txTo), nIn,
+            shrepff, execdata, scriptCode, CTransaction(txTo), nIn,
             sigHashType.withForkValue(newForkValue), Amount::zero(), nullptr,
             SCRIPT_ENABLE_SIGHASH_FORKID | SCRIPT_ENABLE_REPLAY_PROTECTION));
         uint256 manualshrepff;
         BOOST_CHECK(SignatureHash(
-            manualshrepff, scriptCode, CTransaction(txTo), nIn,
+            manualshrepff, execdata, scriptCode, CTransaction(txTo), nIn,
             sigHashType.withForkValue(newForkValue ^ 0xdead), Amount::zero()));
         BOOST_CHECK(shrepff == manualshrepff);
 
         uint256 shrepabcdef;
         BOOST_CHECK(SignatureHash(
-            shrepabcdef, scriptCode, CTransaction(txTo), nIn,
+            shrepabcdef, execdata, scriptCode, CTransaction(txTo), nIn,
             sigHashType.withForkValue(0xabcdef), Amount::zero(), nullptr,
             SCRIPT_ENABLE_SIGHASH_FORKID | SCRIPT_ENABLE_REPLAY_PROTECTION));
         uint256 manualshrepabcdef;
-        BOOST_CHECK(
-            SignatureHash(manualshrepabcdef, scriptCode, CTransaction(txTo), nIn,
-                          sigHashType.withForkValue(0xff1342), Amount::zero()));
+        BOOST_CHECK(SignatureHash(
+            manualshrepabcdef, execdata, scriptCode, CTransaction(txTo), nIn,
+            sigHashType.withForkValue(0xff1342), Amount::zero()));
         BOOST_CHECK(shrepabcdef == manualshrepabcdef);
 
 #if defined(PRINT_SIGHASH_JSON)
@@ -266,21 +270,24 @@ BOOST_AUTO_TEST_CASE(sighash_from_data) {
             BOOST_ERROR("Bad test, couldn't deserialize data: " << strTest);
             continue;
         }
+        const std::optional<ScriptExecutionData> execdata{
+            ScriptExecutionData(scriptCode)};
 
         uint256 shreg;
-        BOOST_CHECK(SignatureHash(shreg, scriptCode, *tx, nIn, sigHashType,
-                                  Amount::zero()));
+        BOOST_CHECK(SignatureHash(shreg, execdata, scriptCode, *tx, nIn,
+                                  sigHashType, Amount::zero()));
         BOOST_CHECK_MESSAGE(shreg.GetHex() == sigHashRegHex, strTest);
 
         uint256 shold;
-        BOOST_CHECK(SignatureHash(shold, scriptCode, *tx, nIn, sigHashType,
-                                  Amount::zero(), nullptr, 0));
+        BOOST_CHECK(SignatureHash(shold, execdata, scriptCode, *tx, nIn,
+                                  sigHashType, Amount::zero(), nullptr, 0));
         BOOST_CHECK_MESSAGE(shold.GetHex() == sigHashOldHex, strTest);
 
         uint256 shrep;
-        BOOST_CHECK(SignatureHash(
-            shrep, scriptCode, *tx, nIn, sigHashType, Amount::zero(), nullptr,
-            SCRIPT_ENABLE_SIGHASH_FORKID | SCRIPT_ENABLE_REPLAY_PROTECTION));
+        BOOST_CHECK(SignatureHash(shrep, execdata, scriptCode, *tx, nIn,
+                                  sigHashType, Amount::zero(), nullptr,
+                                  SCRIPT_ENABLE_SIGHASH_FORKID |
+                                      SCRIPT_ENABLE_REPLAY_PROTECTION));
         BOOST_CHECK_MESSAGE(shrep.GetHex() == sigHashRepHex, strTest);
     }
 }
diff --git a/src/test/txvalidationcache_tests.cpp b/src/test/txvalidationcache_tests.cpp
index 17a13e812..4819f3f4e 100644
--- a/src/test/txvalidationcache_tests.cpp
+++ b/src/test/txvalidationcache_tests.cpp
@@ -52,9 +52,10 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, TestChain100Setup) {
         // Sign:
         std::vector<uint8_t> vchSig;
         uint256 hash;
-        BOOST_CHECK(SignatureHash(hash, scriptPubKey, CTransaction(spends[i]),
-                                  0, SigHashType().withForkId(),
-                                  m_coinbase_txns[0]->vout[0].nValue));
+        BOOST_CHECK(SignatureHash(
+            hash, std::optional(ScriptExecutionData(scriptPubKey)),
+            scriptPubKey, CTransaction(spends[i]), 0,
+            SigHashType().withForkId(), m_coinbase_txns[0]->vout[0].nValue));
         BOOST_CHECK(coinbaseKey.SignECDSA(hash, vchSig));
         vchSig.push_back(uint8_t(SIGHASH_ALL | SIGHASH_FORKID));
         spends[i].vin[0].scriptSig << vchSig;
@@ -218,7 +219,9 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, TestChain100Setup) {
         std::vector<uint8_t> fundingVchSig;
         uint256 fundingSigHash;
         BOOST_CHECK(SignatureHash(
-            fundingSigHash, p2pk_scriptPubKey, CTransaction(funding_tx), 0,
+            fundingSigHash,
+            std::optional(ScriptExecutionData(p2pk_scriptPubKey)),
+            p2pk_scriptPubKey, CTransaction(funding_tx), 0,
             SigHashType().withForkId(), m_coinbase_txns[0]->vout[0].nValue));
         BOOST_CHECK(coinbaseKey.SignECDSA(fundingSigHash, fundingVchSig));
         fundingVchSig.push_back(uint8_t(SIGHASH_ALL | SIGHASH_FORKID));
@@ -346,10 +349,11 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, TestChain100Setup) {
         // Sign
         std::vector<uint8_t> vchSig;
         uint256 hash;
-        BOOST_CHECK(SignatureHash(hash, spend_tx.vout[1].scriptPubKey,
-                                  CTransaction(invalid_with_cltv_tx), 0,
-                                  SigHashType().withForkId(),
-                                  spend_tx.vout[1].nValue));
+        BOOST_CHECK(SignatureHash(
+            hash,
+            std::optional(ScriptExecutionData(spend_tx.vout[1].scriptPubKey)),
+            spend_tx.vout[1].scriptPubKey, CTransaction(invalid_with_cltv_tx),
+            0, SigHashType().withForkId(), spend_tx.vout[1].nValue));
         BOOST_CHECK(coinbaseKey.SignECDSA(hash, vchSig));
         vchSig.push_back(uint8_t(SIGHASH_ALL | SIGHASH_FORKID));
         invalid_with_cltv_tx.vin[0].scriptSig = CScript() << vchSig << 101;
@@ -392,10 +396,11 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, TestChain100Setup) {
         // Sign
         std::vector<uint8_t> vchSig;
         uint256 hash;
-        BOOST_CHECK(SignatureHash(hash, spend_tx.vout[2].scriptPubKey,
-                                  CTransaction(invalid_with_csv_tx), 0,
-                                  SigHashType().withForkId(),
-                                  spend_tx.vout[2].nValue));
+        BOOST_CHECK(SignatureHash(
+            hash,
+            std::optional(ScriptExecutionData(spend_tx.vout[2].scriptPubKey)),
+            spend_tx.vout[2].scriptPubKey, CTransaction(invalid_with_csv_tx), 0,
+            SigHashType().withForkId(), spend_tx.vout[2].nValue));
         BOOST_CHECK(coinbaseKey.SignECDSA(hash, vchSig));
         vchSig.push_back(uint8_t(SIGHASH_ALL | SIGHASH_FORKID));
         invalid_with_csv_tx.vin[0].scriptSig = CScript() << vchSig << 101;
-- 
2.39.5 (Apple Git-154)

